File: editors/nvim/rplugin/python3/denite/filter/converter/devicons.py
"""Converter to prepend devicons to denite candidates."""
from pathlib import Path

from denite.base.filter import Base

DEFAULT_ICON = ''
DIR_ICON = ''
FILENAMES = {
    '.bash_profile': '',
    '.bashrc': '',
    '.ds_store': '',
    '.gitconfig': '',
    '.gitignore': '',
    '.gitlab-ci.yml': '',
    '.gvimrc': '',
    '.vimrc': '',
    '.zshrc': '',
    '_vimrc': '',
    '_gvimrc': '',
    'dockerfile': '',
    'docker-compose.yml': '',
    'dropbox': '',
    'favicon.ico': '',
    'gruntfile.coffee': '',
    'gruntfile.js': '',
    'gruntfile.ls': '',
    'gulpfile.coffee': '',
    'gulpfile.js': '',
    'gulpfile.ls': '',
    'license': '',
    'mix.lock': '',
    'node_modules': '',
    'procfile': '',
    'react.jsx': '',
    'requirements.txt': '',
}
EXTENSIONS = {
    '.ai': '',
    '.awk': '',
    '.bash': '',
    '.bat': '',
    '.bmp': '',
    '.c': '',
    '.cc': '',
    '.clj': '',
    '.cljc': '',
    '.cljs': '',
    '.coffee': '',
    '.conf': '',
    '.cp': '',
    '.cpp': '',
    '.cs': '',
    '.csh': '',
    '.css': '',
    '.cxx': '',
    '.c++': '',
    '.d': '',
    '.dart': '',
    '.db': '',
    '.diff': '',
    '.dump': '',
    '.edn': '',
    '.eex': '',
    '.ejs': '',
    '.elm': '',
    '.erl': '',
    '.ex': '',
    '.exs': '',
    '.f#': '',
    '.fish': '',
    '.fs': '',
    '.fsi': '',
    '.fsscript': '',
    '.fsx': '',
    '.gif': '',
    '.go': '',
    '.h': '',
    '.hbs': '',
    '.hh': '',
    '.hpp': '',
    '.hrl': '',
    '.hs': '',
    '.htm': '',
    '.html': '',
    '.hxx': '',
    '.ico': '',
    '.ini': '',
    '.java': '',
    '.jl': '',
    '.jpeg': '',
    '.jpg': '',
    '.js': '',
    '.json': '',
    '.jsx': '',
    '.ksh': '',
    '.leex': '',
    '.less': '',
    '.lhs': '',
    '.lua': '',
    '.markdown': '',
    '.md': '',
    '.mdx': '',
    '.mjs': '',
    '.ml': 'λ',
    '.mli': 'λ',
    '.mustache': '',
    '.pdf': '',
    '.php': '',
    '.pl': '',
    '.pm': '',
    '.png': '',
    '.pp': '',
    '.ps1': '',
    '.psb': '',
    '.psd': '',
    '.py': '',
    '.pyc': '',
    '.pyd': '',
    '.pyo': '',
    '.r': 'ﳒ',
    '.R': 'ﳒ',
    '.rb': '',
    '.rlib': '',
    '.rmd': '',
    '.rs': '',
    '.rss': '',
    '.sass': '',
    '.scala': '',
    '.scss': '',
    '.sh': '',
    '.slim': '',
    '.sln': '',
    '.sql': '',
    '.styl': '',
    '.suo': '',
    '.swift': '',
    '.t': '',
    '.tex': '',
    '.toml': '',
    '.ts': '',
    '.tsx': '',
    '.txt': '',
    '.twig': '',
    '.vim': '',
    '.vue': '﵂',
    '.xcplayground': '',
    '.xul': '',
    '.yaml': '',
    '.yml': '',
    '.zsh': '',
}


class Filter(Base):
    """Prepend devicons to denite candidate files."""

    def __init__(self, vim):
        super().__init__(vim)
        self.name = 'converter/devicons'
        self.description = 'Prepend devicons to denite candidate files'

    def _get_icon(self, filename):
        filename = Path(filename)
        if filename.is_dir():
            return DIR_ICON

        basefile = filename.stem.lower()
        extension = filename.suffix
        if basefile in FILENAMES:
            return FILENAMES[basefile]
        elif extension in EXTENSIONS:
            return EXTENSIONS[extension]
        else:
            return DEFAULT_ICON

    def filter(self, context):
        """Parse candidates and prepend devicons."""
        for i, candidate in enumerate(context['candidates']):
            if i > 25:
                break

            if 'word' in candidate and 'action__path' in candidate:
                filename = candidate['action__path']
                icon = self._get_icon(filename)
                if icon not in candidate.get('abbr', '')[:10]:
                    candidate[
                        'abbr'
                    ] = f" {icon} {candidate.get('abbr', candidate['word'])}"

        return context['candidates']

File: editors/nvim/lua/devicons.lua
local util = require("utils")
local M = {}

local function get_os_icon()
  local icon = ""
  if util.os.is_windows then
    icon = ""
  elseif util.os.is_mac then
    icon = ""
  else
    -- unix
    if vim.fn.executable("lsb_release") then
      local lsb = util.os.capture("lsb_release -i")
      if string.find(lsb, "Arch", 1, true) then
        icon = ""
      elseif string.find(lsb, "Ubuntu", 1, true) then
        icon = ""
      elseif string.find(lsb, "Cent", 1, true) then
        icon = ""
      elseif string.find(lsb, "Debian", 1, true) then
        icon = ""
      elseif string.find(lsb, "Dock", 1, true) then
        icon = ""
      end
    end
  end
  return icon
end

local os_icon = get_os_icon()

local default_icon = ''

local filetype_icons = {
  -- Exact Match
  ['gruntfile.coffee']   = '',
  ['gruntfile.js']       = '',
  ['gruntfile.ls']       = '',
  ['gulpfile.coffee']    = '',
  ['gulpfile.js']        = '',
  ['gulpfile.ls']        = '',
  ['mix.lock']           = '',
  ['dropbox']            = '',
  ['.ds_store']          = '',
  ['.gitconfig']         = '',
  ['.gitignore']         = '',
  ['.gitlab-ci.yml']     = '',
  ['.bashrc']            = '',
  ['.zshrc']             = '',
  ['.vimrc']             = '',
  ['.gvimrc']            = '',
  ['_vimrc']             = '',
  ['_gvimrc']            = '',
  ['.bashprofile']       = '',
  ['favicon.ico']        = '',
  ['license']            = '',
  ['node_modules']       = '',
  ['react.jsx']          = '',
  ['procfile']           = '',
  ['dockerfile']         = '',
  ['docker-compose.yml'] = '',
  -- More exact match
  ['requirements.txt']   = '',
  -- Extension
  ['styl']               = '',
  ['sass']               = '',
  ['scss']               = '',
  ['htm']                = '',
  ['html']               = '',
  ['slim']               = '',
  ['ejs']                = '',
  ['css']                = '',
  ['less']               = '',
  ['md']                 = '',
  ['mdx']                = '',
  ['markdown']           = '',
  ['rmd']                = '',
  ['json']               = '',
  ['js']                 = '',
  ['mjs']                = '',
  ['jsx']                = '',
  ['rb']                 = '',
  ['php']                = '',
  ['py']                 = '',
  ['pyc']                = '',
  ['pyo']                = '',
  ['pyd']                = '',
  ['coffee']             = '',
  ['mustache']           = '',
  ['hbs']                = '',
  ['conf']               = '',
  ['ini']                = '',
  ['yml']                = '',
  ['yaml']               = '',
  ['toml']               = '',
  ['bat']                = '',
  ['jpg']                = '',
  ['jpeg']               = '',
  ['bmp']                = '',
  ['png']                = '',
  ['gif']                = '',
  ['ico']                = '',
  ['twig']               = '',
  ['cpp']                = '',
  ['c++']                = '',
  ['cxx']                = '',
  ['cc']                 = '',
  ['cp']                 = '',
  ['c']                  = '',
  ['cs']                 = '',
  ['h']                  = '',
  ['hh']                 = '',
  ['hpp']                = '',
  ['hxx']                = '',
  ['hs']                 = '',
  ['lhs']                = '',
  ['lua']                = '',
  ['java']               = '',
  ['sh']                 = '',
  ['fish']               = '',
  ['bash']               = '',
  ['zsh']                = '',
  ['ksh']                = '',
  ['csh']                = '',
  ['awk']                = '',
  ['ps1']                = '',
  ['ml']                 = 'λ',
  ['mli']                = 'λ',
  ['diff']               = '',
  ['db']                 = '',
  ['sql']                = '',
  ['dump']               = '',
  ['clj']                = '',
  ['cljc']               = '',
  ['cljs']               = '',
  ['edn']                = '',
  ['scala']              = '',
  ['go']                 = '',
  ['dart']               = '',
  ['xul']                = '',
  ['sln']                = '',
  ['suo']                = '',
  ['pl']                 = '',
  ['pm']                 = '',
  ['t']                  = '',
  ['rss']                = '',
  ['f#']                 = '',
  ['fsscript']           = '',
  ['fsx']                = '',
  ['fs']                 = '',
  ['fsi']                = '',
  ['rs']                 = '',
  ['rlib']               = '',
  ['d']                  = '',
  ['erl']                = '',
  ['hrl']                = '',
  ['ex']                 = '',
  ['exs']                = '',
  ['eex']                = '',
  ['leex']               = '',
  ['vim']                = '',
  ['ai']                 = '',
  ['psd']                = '',
  ['psb']                = '',
  ['ts']                 = '',
  ['tsx']                = '',
  ['jl']                 = '',
  ['pp']                 = '',
  ['vue']                = '﵂',
  ['elm']                = '',
  ['swift']              = '',
  ['xcplayground']       = '',
  -- more extensions
  ['pdf']                = '',
  ['txt']                = '',
}


M.filetype_icons = setmetatable(filetype_icons,{
    __index = function(filetype_icons, key)
      -- get extension
      local ext = key:match("^.+%.(.+)$")
      if ext then
        return filetype_icons[ext]
      end
      return default_icon
    end
  })

M.fileformat_icons = {
  ['dos']  = '',
  ['mac']  = '',
  ['unix'] = os_icon,
}

return M

